synthesis.go

```diff
@@ -0,0 +1,262 @@
+	// Insert codons
+	weightTable := make(map[string]int)
+	codonInsert := `INSERT INTO codon(codon, aa) VALUES (?, ?)`
+	for _, aminoAcid := range codontable.AminoAcids {
+		for _, codon := range aminoAcid.Codons {
+			db.MustExec(codonInsert, codon.Triplet, aminoAcid.Letter)
+			weightTable[codon.Triplet] = codon.Weight
+
+			codonBias := strings.Count(codon.Triplet, "G") + strings.Count(codon.Triplet, "C")
+			for _, toCodon := range aminoAcid.Codons {
+				if codon.Triplet != toCodon.Triplet {
+					toCodonBias := strings.Count(toCodon.Triplet, "G") + strings.Count(toCodon.Triplet, "C")
+					switch {
+					case codonBias == toCodonBias:
+						db.MustExec(`INSERT INTO codonbias(fromcodon, tocodon, gcbias) VALUES (?, ?, ?)`, codon.Triplet, toCodon.Triplet, "NA")
+					case codonBias > toCodonBias:
+						db.MustExec(`INSERT INTO codonbias(fromcodon, tocodon, gcbias) VALUES (?, ?, ?)`, codon.Triplet, toCodon.Triplet, "AT")
+					case codonBias < toCodonBias:
+						db.MustExec(`INSERT INTO codonbias(fromcodon, tocodon, gcbias) VALUES (?, ?, ?)`, codon.Triplet, toCodon.Triplet, "GC")
+					}
+				}
+			}
+		}
+	}
+
+	// Insert seq and history
+	pos := 0
+	for i := 0; i < len(sequence)-3; i = i + 3 {
+		codon := sequence[i : i+3]
+		db.MustExec(`INSERT INTO seq(pos) VALUES (?)`, pos)
+		db.MustExec(`INSERT INTO history(pos, codon, step) VALUES (?, ?, 0)`, pos, codon)
+		db.MustExec(`INSERT INTO weights(pos, codon, weight) VALUES (?,?,?)`, pos, codon, weightTable[codon])
+		pos++
+	}
+
+	var err error
+	// For a maximum of 100 iterations, see if we can do better
+	for i := 1; i < 100; i++ {
+		suggestions := findProblems(sequence, problematicSequenceFuncs)
+		// If there are no suggestions, break the iteration!
+		if len(suggestions) == 0 {
+			break
+		}
+		for _, suggestion := range suggestions { // if you want to add overlaps, add suggestionIndex
+			// First, let's insert the suggestions that we found using our problematicSequenceFuncs
+			_, err = db.Exec(`INSERT INTO suggestedfix(step, start, end, gcbias, quantityfixes, suggestiontype) VALUES (?, ?, ?, ?, ?, ?)`, i, suggestion.Start, suggestion.End, suggestion.Bias, suggestion.QuantityFixes, suggestion.SuggestionType)
+			if err != nil {
+				return sequence, err
+			}
+		}
+
+		// The following statements are the magic sauce that makes this all worthwhile.
+		// Parameters: step, gcbias, start, end, quantityfix
+		sqlFix1 := `INSERT INTO history
+		            (codon,
+		             pos,
+		             step)
+		SELECT t.codon,
+		       t.pos,
+		       ? AS step
+		FROM   (SELECT cb.tocodon AS codon,
+		               s.pos      AS pos
+		        FROM   seq AS s
+		               JOIN history AS h
+		                 ON h.pos = s.pos
+		               JOIN weights AS w
+		                 ON w.pos = s.pos
+		               JOIN codon AS c
+		                 ON h.codon = c.codon
+		               JOIN codonbias AS cb
+		                 ON cb.fromcodon = c.codon
+		        WHERE `
```

Comment:
Is there nothing like python pandas for golang?

relying on sql, even if in memory, for stuff like this seems heavy, especially when computations get big.

Reply:
I use SQL here as an embedded declarative language that people don't have to learn a domain specific language to use. 

Pandas wouldn't really work here, since I'm explicitly looking at doing joins from a variety of sources. The closest thing would be an embedded thing like - https://pkg.go.dev/modernc.org/ql?utm_source=godoc - but SQLite is so widely used, I thought might as well go for something where you won't have to learn an entirely new thing to use.

Reply:
Computations stay sub-second, which is still 100x better than alternatives out there. They're pretty bounded by the size of a given gene.

Reply:
sounds good. Also our convo in discord made a lot of sense.